### PR TITLE
Add tests for scissors

### LINKED ISSUE #3871: Tests for Scissors
Particularly we need tests for zero size, but general tests would be good too.

### PR DESCRIPTION
**Pull request by artyomd**
**Checklist**

- [x] Run `cargo clippy`.
- [x] Run `cargo clippy --target wasm32-unknown-unknown` if applicable.
- [ ] ~~Add change to CHANGELOG.md. See simple instructions inside file.~~ (no need)

**Description**
Added some tests for scissors. Resolves #3871

### CODE REVIEW

File: tests/Cargo.toml [hunk 1]
```diff
@@ -27,6 +27,7 @@ pollster.workspace = true
 png.workspace = true
 wgpu.workspace = true
 wgt.workspace = true
+bytemuck.workspace = true
```
Comment (cwfitzgerald) [inline tests/Cargo.toml]: Mind alphabetical order please

File: tests/src/image.rs [hunk 2]
```diff
@@ -526,3 +527,51 @@ impl ReadbackBuffers {
                 .unwrap_or(true)
     }
 }
+
+pub fn capture_rgba_u8_texture(
```
Comment (cwfitzgerald) [inline tests/src/image.rs]: Couldn't you use the already existing ReadbackBuffers to do this (and the SPI tests should use them too?)

File: tests/tests/scissor_tests/mod.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    width: u32,
+    height: u32,
+}
+
+fn scissor_test_impl(ctx: &TestingContext, scissor_rect: Rect, expected_data: [u8; 16]) {
+    let vertex_data: [f32; 12] = [
+        -1.0_f32, -1.0_f32, 0.0_f32, 1.0_f32, -1.0_f32, 0.0_f32, 1.0_f32, 1.0_f32, 0.0_f32,
+        -1.0_f32, 1.0_f32, 0.0_f32,
+    ];
+    let index_data: [i16; 6] = [0, 1, 3, 1, 3, 2];
```
Comment (cwfitzgerald) [inline tests/tests/scissor_tests/mod.rs]: If you're just trying to make fullscreen geometry, you can use the same VS as mipmap does here: https://github.com/gfx-rs/wgpu/blob/trunk/examples/mipmap/src/blit.wgsl

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Looking good minus comments

File: tests/Cargo.toml (hunk 1 above)
Comment (artyomd) [reply tests/Cargo.toml]: 👍 I fixed the order of all the dependencies in this file

File: tests/src/image.rs (hunk 2 above)
Comment (artyomd) [reply tests/src/image.rs]: This is a bit tricky. The entire problem is the COPY_BYTES_PER_ROW_ALIGNMENT requirement. Both in the scissors test and in the spi, the tests render the 2x2 texture. At first, I thought that it will be a good idea to support copying textures that do not meet the requirement in the readback buffers implementation, but when I started implementing that, at some point it felt that readback buffers needed their own tests as the implementation was getting too complicated (multiple types of textures, multiple ways to copy them, etc..) Personally, I like to keep tests as simple as possible, so I decided the best approach to this will be just to rewrite tests to meet the alignment requirement. What do you think about this?

File: tests/tests/scissor_tests/mod.rs (hunk 3 above)
Comment (artyomd) [reply tests/tests/scissor_tests/mod.rs]: Thank you for mentioning that. I wasn't familiar with that technique. Utilizing it allowed me to simplify the test a lot.

File: tests/src/image.rs (hunk 2 above)
Comment (cwfitzgerald) [reply tests/src/image.rs]: Shouldn't you be able to manage the COPY_BYTES_PER_ROW_ALIGNMENT by purely byte manipulation?

When computing the buffer size and executing the copy round up the `bytes_per_row` to `COPY_BYTES_PER_ROW_ALIGNMENT`. Then using the real bytes_per_row information, copy only the chunks that have real data in them into the vector.

I would prefer to keep the tests small, as it helps with understanding what exactly is getting tested.

Comment (artyomd) [reply tests/src/image.rs]: "I would prefer to keep the tests small, as it helps with understanding what exactly is getting tested."
I agree with this.

"Shouldn't you be able to manage the COPY_BYTES_PER_ROW_ALIGNMENT by purely byte manipulation?"
Yes, and that was done in the original `capture_rgba_u8_texture` function that was written in the SPI test.

I think that I may have overanalyzed your initial comment. Just to make it clear, you want to keep the `capture_rgba_u8_texture` function in the image.rs, but use ReadbackBuffers in the implementation, am I right?

Comment (cwfitzgerald) [reply tests/src/image.rs]: Ideally ReadbackBuffers can just deal with the COPY_BYTES_PER_ROW requirement internally. Then all the tests can just use ReadbackBuffers without thinking about it.

Comment (artyomd) [reply tests/src/image.rs]: Done.

File: tests/src/image.rs [hunk 4]
```diff
@@ -1,6 +1,8 @@
 use std::{borrow::Cow, ffi::OsStr, io, path::Path};
+
 use wgpu::util::DeviceExt;
 use wgpu::*;
+use wgt::math::align_to;
```
Comment (cwfitzgerald) [inline tests/src/image.rs]: We shouldn't dep on wgt if we can help it, this is re-exported in wgpu.

File: tests/src/image.rs [hunk 5]
```diff
@@ -458,20 +472,37 @@ pub struct ReadbackBuffers {
 impl ReadbackBuffers {
     pub fn new(device: &Device, texture: &Texture) -> Self {
         let (block_width, block_height) = texture.format().block_dimensions();
-        let base_size = (texture.width() / block_width)
-            * (texture.height() / block_height)
-            * texture.depth_or_array_layers();
+        let should_align_buffer_size = !([
+            TextureFormat::Depth24Plus,
+            TextureFormat::Depth24PlusStencil8,
```
Comment (cwfitzgerald) [inline tests/src/image.rs:476]: I don't understand why we're only aligning to the buffer size on these types. No matter the texture format, we need to align the copies to the bytes_per_row alignment.

File: tests/src/image.rs [hunk 6]
```diff
@@ -505,24 +550,70 @@ impl ReadbackBuffers {
...
+            * self.texture_format.block_size(aspect).unwrap_or(4);
+        let expected_buffer_size = expected_bytes_per_row
+            * (self.texture_height / block_height)
+            * self.texture_depth_or_array_layers;
+        let data: BufferView = buffer_slice.get_mapped_range();
+        if expected_buffer_size as usize == data.len() {
+            data.to_vec()
+        } else {
+            bytemuck::cast_slice(&data)
+                .to_vec()
```
Comment (cwfitzgerald) [inline tests/src/image.rs]: This to_vec is unneeded

File: tests/src/image.rs [hunk 7]
```diff
@@ -505,24 +550,70 @@ impl ReadbackBuffers {
...
 
-        is_zero(device, &self.buffer)
+        is_zero(device, &self.buffer, self.buffer_aspect())
             && self
                 .buffer_stencil
                 .as_ref()
-                .map(|buffer_stencil| is_zero(device, buffer_stencil))
+                .map(|buffer_stencil| {
+                    is_zero(device, buffer_stencil, Some(TextureAspect::StencilOnly))
+                })
```
Comment (cwfitzgerald) [inline tests/src/image.rs]: Could we split this into two bools which are then and'ed

File: tests/tests/shader_primitive_index/mod.rs [hunk 8]
```diff
@@ -168,83 +169,31 @@ fn pulling_common(
...
-                rows_per_image: None,
-            },
-        },
-        texture_size,
-    );
-
+        rpass.set_pipeline(&pipeline);
+        rpass.set_index_buffer(index_buffer.slice(..), wgpu::IndexFormat::Uint32);
+        rpass.set_vertex_buffer(0, vertex_buffer.slice(..));
+        function(&mut rpass);
```
Comment (cwfitzgerald) [inline tests/tests/shader_primitive_index/mod.rs]: What is this call?

**Review by cwfitzgerald — CHANGES_REQUESTED** [review]
Much better direction, some comments

File: tests/src/image.rs (hunk 4 above)
Comment (artyomd) [reply tests/src/image.rs]: fixed

File: tests/src/image.rs (hunk 5 above)
Comment (artyomd) [reply tests/src/image.rs:476]: It is the opposite, we are **not aligning** if the texture format is one of these. The reason why we are not checking the alignment is that we are using a custom compute pipeline to copy the data if the depth format is Depth24Plus. And as far as I know, that is because as of [webgpu spec](https://gpuweb.github.io/gpuweb/#depth-formats), depth aspects of textures with those formats are an invalid copy source.

File: tests/src/image.rs (hunk 6 above)
Comment (artyomd) [reply tests/src/image.rs]: removed

File: tests/src/image.rs (hunk 7 above)
Comment (artyomd) [reply tests/src/image.rs]: Honestly, I prefer the way that it was written, but I agree that at first, it is a little bit hard to grasp this logic.
refactored this one, too

File: tests/tests/shader_primitive_index/mod.rs (hunk 8 above)
Comment (artyomd) [reply tests/tests/shader_primitive_index/mod.rs]: So this is the original name of the parameter that was written. It was very descriptive,  but as I was refactoring this function, somehow I didn't notice that.

File: tests/src/image.rs (hunk 5 above)
Comment (artyomd) [reply tests/src/image.rs:476]: Improved the readability of these lines too

**Review by cwfitzgerald — APPROVED** [review]
Amazing
